pkg/serviceimport/map.go

```diff
@@ -66,6 +59,37 @@ func (m *Map) SelectIP(namespace, name string, checkCluster func(string) bool) (
 	return "", true
 }
 
+func (m *Map) GetIPs(namespace, name string, checkCluster func(string) bool) ([]string, bool) {
+	var svcInfo *serviceInfo
+	var ok bool
+
+	serviceIps := make([]string, 0)
+
+	m.RLock()
+	defer m.RUnlock()
+
+	if svcInfo, ok = m.svcMap[keyFunc(namespace, name)]; !ok {
+		return nil, false
+	}
+
+	if !svcInfo.isHeadless {
+		ip, found := m.selectIP(svcInfo, checkCluster)
+		if ip != "" {
+			serviceIps = append(serviceIps, ip)
+		}
+
+		return serviceIps, found
+	}
+
+	for cluster, ips := range svcInfo.clusterIPs {
+		if checkCluster(cluster) && len(ips) > 0 {
+			serviceIps = append(serviceIps, ips...)
+		}
+	}
```

Comment:
The prior code strived to release the lock ASAP and capture all fields to access in the critical section, as `queue` et al are recreated on change - we should do he same here:

```
clusterIPs, queue, counter, isHeadless := func() (map[string][]string, []clusterInfo, *uint64, bool) {
    m.RLock()
    defer m.RUnlock()

    si, ok := m.svcMap[keyFunc(namespace, name)]
    if !ok {
        return nil, nil, nil, false
    }

    return si.clusterIPs , si.clustersQueue, &si.rrCount, si.isHeadless
}()

if clusterIPs == nil {
    return nil, false
}

if !isHeadless {
    ip := m.selectIP(queue, counter, checkCluster)
    ...
}

for cluster, ips := range clusterIPs {
    ...
}
```

Reply:
Will do.